src/src/PluginStructs/P104_data_struct.cpp

```diff
@@ -0,0 +1,2458 @@
+#include "../PluginStructs/P104_data_struct.h"
+
+#ifdef USES_P104
+
+# include "../Helpers/ESPEasy_Storage.h"
+# include "../Helpers/Numerical.h"
+# include "../WebServer/Markup_Forms.h"
+# include "../WebServer/WebServer.h"
+# include "../WebServer/Markup.h"
+# include "../WebServer/HTML_wrappers.h"
+# include "../ESPEasyCore/ESPEasyRules.h"
+# include "../Globals/ESPEasy_time.h"
+# include "../Globals/RTC.h"
+
+# include <vector>
+# include <MD_Parola.h>
+# include <MD_MAX72xx.h>
+
+// Needed also here for PlatformIO's library finder as the .h file
+// is in a directory which is excluded in the src_filter
+
+# if defined(P104_USE_NUMERIC_DOUBLEHEIGHT_FONT) || defined(P104_USE_FULL_DOUBLEHEIGHT_FONT)
+void createHString(String& string); // Forward definition
+# endif // if defined(P104_USE_NUMERIC_DOUBLEHEIGHT_FONT) || defined(P104_USE_FULL_DOUBLEHEIGHT_FONT)
+void reverseStr(String& str);       // Forward definition
+
+/****************************************************************
+ * Constructor
+ ***************************************************************/
+P104_data_struct::P104_data_struct(MD_MAX72XX::moduleType_t _mod,
+                                   taskIndex_t              _taskIndex,
+                                   int8_t                   _cs_pin,
+                                   uint8_t                  _modules,
+                                   uint8_t                  _zonesCount)
+  : mod(_mod), taskIndex(_taskIndex), cs_pin(_cs_pin), modules(_modules), expectedZones(_zonesCount) {
+  if (Settings.InitSPI > 0) { // FIXME tonhuisman: Replace by isSPI_valid() once merged from PR #3773
+    P = new MD_Parola(mod, cs_pin, modules);
```

Comment:
This is allocated, but never deleted -> memory leak.
Better have all pointers like these deleted in the destructor (now there isn't one).
The most fool proof solution for this is to make it not a standard pointer, but one of those autopointer types.
Those will destruct to where they point as soon as the autopointer object gets out of scope.

autopointer itself is deprecated, but there are several types which can be used which have similar properties.

Reply:
Added a destructor.

Reply:
The Parola library doesn't have a virtual destructor, and when I make that virtual, the ESP reboots when the `P` object is `delete`ed 😨 I've documented that in the source.
Now the pointer is set to `nullptr` (and the same for the `pM` object when used).

NB: Memory _is_ freed properly once the plugin is disabled (according to the numbers reported by the watchdog).

Reply:
> NB: Memory _is_ freed properly once the plugin is disabled (according to the numbers reported by the watchdog).

Hmm that's strange.
Have to look at the code then to see what is so special about that class.